smslibrary/src/main/java/com/eis/smslibrary/background/SMSKeepAliveService.java

```diff
@@ -0,0 +1,52 @@
+package com.eis.smslibrary.background;
+
+import android.app.Service;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.IBinder;
+
+import com.eis.smslibrary.SMSReceivedBroadcastReceiver;
+
+/**
+ * Android service that is kept alive and has the BroadcastReceiver attached so that it doesn't die on app closed
+ *
+ * @author Marco Tommasini & Luca Crema
+ */
+public final class SMSKeepAliveService extends Service {
+
+    private static SMSReceivedBroadcastReceiver broadcastReceiverInstance;
+
+    @Override
+    public IBinder onBind(Intent arg0) {
+        return null;
+    }
+
+    /**
+     * On startup it registers the broadcastReceiver
+     */
+    @Override
+    public void onCreate() {
+        registerMessageBroadcastReceiver();
+    }
+
+    /**
+     * When the service is destroyed it un-registers the broadcastReceiver
+     */
+    @Override
+    public void onDestroy() {
+        unregisterReceiver(broadcastReceiverInstance);
+        broadcastReceiverInstance = null;
+    }
+
+    /**
+     * Registers the broadcastReceiver on this service, so that it will also be called
+     * when the app is closed
+     */
+    private void registerMessageBroadcastReceiver() {
```

Comment:
There's one thing I don't understand: we already have a registered SMSReceivedBroadcastReceiver, because it's declared in the manifest of smslibrary, and since it is a BroadcastReceiver it will always be called when a message is received, even if the app was destroyed. Then what's the point of having an SMSKeepAliveService that registers another BroadcastReceiver, when we already have one enabled?

Reply:
Because our tests showed that BroadcastReceiver was not being called on most devices if the app was dead or if the phone was restarted, so as suggested in a few [StackOverflow answers ](https://stackoverflow.com/questions/16824341/keep-broadcast-receiver-running-after-application-is-closed) we decided to create the `SMSKeepAliveService`

Reply:
If you read the documentation I linked you'll see that there's a difference between BroadcastReceivers registered in the manifest and those registered with a Context:
in the first case, they are called even when the app is dead; in the second case, they're only called when their Context is active, and the Stackoverflow answer you posted refers to this second case. Our library's case is the first, so a service is not necessary.

Reply:
What about our tests? We tested with a service to be started on message arrival that had to print a notification, and it wasn't working properly every time: when we closed the app the notification would appear only if the message arrived within a few minutes from when the app was closed.

Reply:
Manifest declared Receivers are supposed to work no matter what happens to the app. During the demonstration of group's 1 demo for the killer app another activity was started in addition to the killer app, that was due to a Receiver activating from a dumb test app of mine, days upon days after installing it and after various cold boots of the emulator. A Receiver registered through a context will die out the moment the activity/application (depends on which you pass) is destroyed. I don't know what kind of tests you guys did but I think we need to delve further into this.

Reply:
I don't know what tests you did, and I'm sorry if I sound rude, but I trust Android's documentation more than you. This afternoon I'll try writing a simple app to receive messages using this library and I'll see what happens.

Reply:
Ok, if the two of you are 100% sure about this I'll remove the service.

Reply:
I'm not 100% sure, I'd like to try the library and fix the BroadcastReceiver registered in the manifest first (if it doesn't work), if that is not possible then I'll accept the service. Even though it's not guaranteed that it will always run in the background since Android can kill it if it consumes too much battery, it's still better than SMS reception not working in the background at all.

Reply:
It seems that in order for the BroadcastReceiver to work it should be declared in the app's manifest, not in the library's manifest

Reply:
This sounds very weird, I thought the manifests got merged while compiling

Reply:
@CremaLuca could you create another branch in your fork starting from Cogno-IDU:master and add there the application you were using to test the BroadcastReceiver?

Reply:
Here's a test app with a BroadcastReceiver that opens the main activity everytime a message is received, as you can see it works even after closing the app and after a reboot: [https://youtu.be/X4wUtfc9-vM](https://youtu.be/X4wUtfc9-vM)

Reply:
@giovannivelludo I removed the KeepAliveService because it should be useless and needs more research, but proving your point on an emulator is not the way I'd do it. We tested on real physical phones and Huawei seems to have a few problems every now and then about the BroadcastReceiver.

I would keep it like this until this gets merged and until I have some more time to test, if this needs the service again I'll do another pull request to add it.

Reply:
Could you point me to the branch in your repo where you developed the app you were testing?